File: bot/cogs/fun.py
import discord
from discord.ext import commands
from discord import app_commands
import aiohttp
import random
from typing import Optional
import config
from bot.utils.embeds import success_embed, error_embed, info_embed


class Fun(commands.Cog):
    def __init__(self, bot):
        self.bot = bot
        self.kawaii_api = "https://kawaii.red/api/gif"
        self.eight_ball_responses = [
            "It is certain.", "Without a doubt.", "Yes, definitely.",
            "You can count on it.", "From what I see, yes.",
            "Very likely.", "Outlook is good.", "Yes.",
            "Signs point to yes.", "Reply hazy, try again.",
            "Ask again later.", "Better not tell you now.",
            "Cannot predict now.", "Concentrate and ask again.",
            "Don't count on it.", "My answer is no.", "My sources say no.",
            "Outlook is not so good.", "Very doubtful."
        ]

    async def get_kawaii_gif(self, endpoint: str, session: aiohttp.ClientSession):
        headers = {}
        if config.KAWAII_API_KEY:
            headers["Authorization"] = f"Bearer {config.KAWAII_API_KEY}"

        async with session.get(f"{self.kawaii_api}/{endpoint}", headers=headers) as resp:
            if resp.status == 200:
                data = await resp.json()
                return data.get("response")
            return None

    @commands.hybrid_command(name="hug", description="Give someone a hug")
    @commands.cooldown(1, 3, commands.BucketType.user)
    async def hug(self, ctx: commands.Context, member: discord.Member):
        if member.id == ctx.author.id:
            await ctx.send(embed=error_embed("You cannot hug yourself"))
            return

        async with aiohttp.ClientSession() as session:
            gif_url = await self.get_kawaii_gif("hug", session)

        embed = discord.Embed(
            description=f"**{ctx.author.mention}** gives **{member.mention}** a hug 🤗",
            color=discord.Color.pink()
        )
        if gif_url:
            embed.set_image(url=gif_url)
        await ctx.send(embed=embed)

    @commands.hybrid_command(name="kiss", description="Kiss someone")
    @commands.cooldown(1, 3, commands.BucketType.user)
    async def kiss(self, ctx: commands.Context, member: discord.Member):
        if member.id == ctx.author.id:
            await ctx.send(embed=error_embed("You cannot kiss yourself"))
            return

        async with aiohttp.ClientSession() as session:
            gif_url = await self.get_kawaii_gif("kiss", session)

        embed = discord.Embed(
            description=f"**{ctx.author.mention}** kisses **{member.mention}** 💋",
            color=discord.Color.red()
        )
        if gif_url:
            embed.set_image(url=gif_url)
        await ctx.send(embed=embed)

    @commands.hybrid_command(name="pat", description="Pat someone's head")
    @commands.cooldown(1, 3, commands.BucketType.user)
    async def pat(self, ctx: commands.Context, member: discord.Member):
        if member.id == ctx.author.id:
            await ctx.send(embed=error_embed("You cannot pat yourself"))
            return

        async with aiohttp.ClientSession() as session:
            gif_url = await self.get_kawaii_gif("pat", session)

        embed = discord.Embed(
            description=f"**{ctx.author.mention}** pats **{member.mention}** ✋",
            color=discord.Color.blue()
        )
        if gif_url:
            embed.set_image(url=gif_url)
        await ctx.send(embed=embed)

    @commands.hybrid_command(name="poke", description="Poke someone")
    @commands.cooldown(1, 3, commands.BucketType.user)
    async def poke(self, ctx: commands.Context, member: discord.Member):
        if member.id == ctx.author.id:
            await ctx.send(embed=error_embed("You cannot poke yourself"))
            return

        async with aiohttp.ClientSession() as session:
            gif_url = await self.get_kawaii_gif("poke", session)

        embed = discord.Embed(
            description=f"**{ctx.author.mention}** pokes **{member.mention}** 👉",
            color=discord.Color.orange()
        )
        if gif_url:
            embed.set_image(url=gif_url)
        await ctx.send(embed=embed)

    @commands.hybrid_command(name="slap", description="Slap someone")
    @commands.cooldown(1, 3, commands.BucketType.user)
    async def slap(self, ctx: commands.Context, member: discord.Member):
        if member.id == ctx.author.id:
            await ctx.send(embed=error_embed("You cannot slap yourself"))
            return

        async with aiohttp.ClientSession() as session:
            gif_url = await self.get_kawaii_gif("slap", session)

        embed = discord.Embed(
            description=f"**{ctx.author.mention}** slaps **{member.mention}** 👋",
            color=discord.Color.red()
        )
        if gif_url:
            embed.set_image(url=gif_url)
        await ctx.send(embed=embed)

    @commands.hybrid_command(name="cuddle", description="Cuddle someone")
    @commands.cooldown(1, 3, commands.BucketType.user)
    async def cuddle(self, ctx: commands.Context, member: discord.Member):
        if member.id == ctx.author.id:
            await ctx.send(embed=error_embed("You cannot cuddle yourself"))
            return

        async with aiohttp.ClientSession() as session:
            gif_url = await self.get_kawaii_gif("cuddle", session)

        embed = discord.Embed(
            description=f"**{ctx.author.mention}** cuddles **{member.mention}** 🥰",
            color=discord.Color.pink()
        )
        if gif_url:
            embed.set_image(url=gif_url)
        await ctx.send(embed=embed)

    @commands.hybrid_command(name="tickle", description="Tickle someone")
    @commands.cooldown(1, 3, commands.BucketType.user)
    async def tickle(self, ctx: commands.Context, member: discord.Member):
        if member.id == ctx.author.id:
            await ctx.send(embed=error_embed("You cannot tickle yourself"))
            return

        async with aiohttp.ClientSession() as session:
            gif_url = await self.get_kawaii_gif("tickle", session)

        embed = discord.Embed(
            description=f"**{ctx.author.mention}** tickles **{member.mention}** 🤭",
            color=discord.Color.gold()
        )
        if gif_url:
            embed.set_image(url=gif_url)
        await ctx.send(embed=embed)

    @commands.hybrid_command(name="bite", description="Bite someone")
    @commands.cooldown(1, 3, commands.BucketType.user)
    async def bite(self, ctx: commands.Context, member: discord.Member):
        if member.id == ctx.author.id:
            await ctx.send(embed=error_embed("You cannot bite yourself"))
            return

        async with aiohttp.ClientSession() as session:
            gif_url = await self.get_kawaii_gif("bite", session)

        embed = discord.Embed(
            description=f"**{ctx.author.mention}** bites **{member.mention}** 🦷",
            color=discord.Color.purple()
        )
        if gif_url:
            embed.set_image(url=gif_url)
        await ctx.send(embed=embed)

    @commands.hybrid_command(name="lick", description="Lick someone")
    @commands.cooldown(1, 3, commands.BucketType.user)
    async def lick(self, ctx: commands.Context, member: discord.Member):
        if member.id == ctx.author.id:
            await ctx.send(embed=error_embed("You cannot lick yourself"))
            return

        async with aiohttp.ClientSession() as session:
            gif_url = await self.get_kawaii_gif("lick", session)

        embed = discord.Embed(
            description=f"**{ctx.author.mention}** licks **{member.mention}** 👅",
            color=discord.Color.magenta()
        )
        if gif_url:
            embed.set_image(url=gif_url)
        await ctx.send(embed=embed)

    @commands.hybrid_command(name="8ball", description="Ask the magic 8-ball a question")
    @commands.cooldown(1, 5, commands.BucketType.user)
    async def eight_ball(self, ctx: commands.Context, *, question: str):
        response = random.choice(self.eight_ball_responses)

        embed = discord.Embed(
            title="🎱 Magic 8-Ball",
            color=discord.Color.blue()
        )
        embed.add_field(name="Question", value=question, inline=False)
        embed.add_field(name="Answer", value=response, inline=False)
        embed.set_footer(text=f"Requested by {ctx.author}", icon_url=ctx.author.display_avatar.url)
        await ctx.send(embed=embed)

    @commands.hybrid_command(name="joke", description="Get a random joke")
    @commands.cooldown(1, 5, commands.BucketType.user)
    async def joke(self, ctx: commands.Context):
        jokes = [
            "Why don't scientists trust atoms? Because they make up everything.",
            "Why did the programmer quit his job? Because he didn't get arrays.",
            "Why did the scarecrow win an award? He was outstanding in his field.",
            "What do you call fake spaghetti? An impasta.",
            "Why did the math book look sad? It had too many problems.",
            "What do you call cheese that isn't yours? Nacho cheese.",
            "Why can't your nose be 12 inches long? Because then it would be a foot.",
            "Why did the bicycle fall over? It was two tired.",
            "What do you call a pile of cats? A meowtain.",
            "Why did the golfer bring two pairs of pants? In case he got a hole in one."
        ]

        joke = random.choice(jokes)

        embed = discord.Embed(
            title="😂 Joke",
            description=joke,
            color=discord.Color.gold()
        )
        embed.set_footer(text=f"Requested by {ctx.author}", icon_url=ctx.author.display_avatar.url)
        await ctx.send(embed=embed)

    @commands.hybrid_command(name="meme", description="Get a random meme")
    @commands.cooldown(1, 5, commands.BucketType.user)
    async def meme(self, ctx: commands.Context):
        async with aiohttp.ClientSession() as session:
            async with session.get("https://meme-api.com/gimme") as resp:
                if resp.status != 200:
                    await ctx.send(embed=error_embed("Unable to fetch a meme"))
                    return

                data = await resp.json()

                embed = discord.Embed(
                    title=data.get("title", "Meme"),
                    url=data.get("postLink", ""),
                    color=discord.Color.random()
                )
                embed.set_image(url=data.get("url"))
                embed.set_footer(text=f"👍 {data.get('ups', 0)} | r/{data.get('subreddit', 'memes')}")
                await ctx.send(embed=embed)


async def setup(bot):
    await bot.add_cog(Fun(bot))

File: src/commands/fun.rs
use crate::handler::{Context, Error};
use crate::utils::embeds::*;
use poise::serenity_prelude as serenity;
use rand::Rng;
use serde::Deserialize;

#[derive(Debug, Deserialize)]
struct KawaiiResponse {
    response: String,
}

#[derive(Debug, Deserialize)]
struct MemeResponse {
    title: String,
    url: String,
    #[serde(rename = "postLink")]
    post_link: String,
    subreddit: String,
    ups: i32,
}

async fn get_kawaii_gif(endpoint: &str, api_key: Option<&String>) -> Result<Option<String>, Error> {
    let client = reqwest::Client::new();
    let mut request = client.get(format!("https://kawaii.red/api/gif/{}", endpoint));

    if let Some(key) = api_key {
        request = request.header("Authorization", format!("Bearer {}", key));
    }

    let response = request.send().await?;
    if response.status().is_success() {
        let data: KawaiiResponse = response.json().await?;
        Ok(Some(data.response))
    } else {
        Ok(None)
    }
}

#[poise::command(slash_command, category = "Fun")]
pub async fn hug(
    ctx: Context<'_>,
    #[description = "User to hug"] member: serenity::User,
) -> Result<(), Error> {
    if member.id == ctx.author().id {
        ctx.send(poise::CreateReply::default().embed(error_embed("You cannot hug yourself")))
            .await?;
        return Ok(());
    }

    let gif_url = get_kawaii_gif("hug", ctx.data().config.kawaii_api_key.as_ref()).await?;

    let mut embed = serenity::CreateEmbed::default()
        .description(format!("**{}** gives **{}** a hug 🤗", ctx.author().name, member.name))
        .color(serenity::Color::from_rgb(255, 192, 203));

    if let Some(url) = gif_url {
        embed = embed.image(url);
    }

    ctx.send(poise::CreateReply::default().embed(embed)).await?;
    Ok(())
}

#[poise::command(slash_command, category = "Fun")]
pub async fn kiss(
    ctx: Context<'_>,
    #[description = "User to kiss"] member: serenity::User,
) -> Result<(), Error> {
    if member.id == ctx.author().id {
        ctx.send(poise::CreateReply::default().embed(error_embed("You cannot kiss yourself")))
            .await?;
        return Ok(());
    }

    let gif_url = get_kawaii_gif("kiss", ctx.data().config.kawaii_api_key.as_ref()).await?;

    let mut embed = serenity::CreateEmbed::default()
        .description(format!("**{}** kisses **{}** 💋", ctx.author().name, member.name))
        .color(serenity::Color::from_rgb(231, 76, 60));

    if let Some(url) = gif_url {
        embed = embed.image(url);
    }

    ctx.send(poise::CreateReply::default().embed(embed)).await?;
    Ok(())
}

#[poise::command(slash_command, category = "Fun")]
pub async fn pat(
    ctx: Context<'_>,
    #[description = "User to pat"] member: serenity::User,
) -> Result<(), Error> {
    if member.id == ctx.author().id {
        ctx.send(poise::CreateReply::default().embed(error_embed("You cannot pat yourself")))
            .await?;
        return Ok(());
    }

    let gif_url = get_kawaii_gif("pat", ctx.data().config.kawaii_api_key.as_ref()).await?;

    let mut embed = serenity::CreateEmbed::default()
        .description(format!("**{}** pats **{}** ✋", ctx.author().name, member.name))
        .color(serenity::Color::from_rgb(52, 152, 219));

    if let Some(url) = gif_url {
        embed = embed.image(url);
    }

    ctx.send(poise::CreateReply::default().embed(embed)).await?;
    Ok(())
}

#[poise::command(slash_command, category = "Fun")]
pub async fn slap(
    ctx: Context<'_>,
    #[description = "User to slap"] member: serenity::User,
) -> Result<(), Error> {
    if member.id == ctx.author().id {
        ctx.send(poise::CreateReply::default().embed(error_embed("You cannot slap yourself")))
            .await?;
        return Ok(());
    }

    let gif_url = get_kawaii_gif("slap", ctx.data().config.kawaii_api_key.as_ref()).await?;

    let mut embed = serenity::CreateEmbed::default()
        .description(format!("**{}** slaps **{}** 👋", ctx.author().name, member.name))
        .color(serenity::Color::from_rgb(231, 76, 60));

    if let Some(url) = gif_url {
        embed = embed.image(url);
    }

    ctx.send(poise::CreateReply::default().embed(embed)).await?;
    Ok(())
}

#[poise::command(slash_command, category = "Fun")]
pub async fn cuddle(
    ctx: Context<'_>,
    #[description = "User to cuddle"] member: serenity::User,
) -> Result<(), Error> {
    if member.id == ctx.author().id {
        ctx.send(poise::CreateReply::default().embed(error_embed("You cannot cuddle yourself")))
            .await?;
        return Ok(());
    }

    let gif_url = get_kawaii_gif("cuddle", ctx.data().config.kawaii_api_key.as_ref()).await?;

    let mut embed = serenity::CreateEmbed::default()
        .description(format!("**{}** cuddles **{}** 🥰", ctx.author().name, member.name))
        .color(serenity::Color::from_rgb(255, 192, 203));

    if let Some(url) = gif_url {
        embed = embed.image(url);
    }

    ctx.send(poise::CreateReply::default().embed(embed)).await?;
    Ok(())
}

#[poise::command(slash_command, category = "Fun")]
pub async fn poke(
    ctx: Context<'_>,
    #[description = "User to poke"] member: serenity::User,
) -> Result<(), Error> {
    if member.id == ctx.author().id {
        ctx.send(poise::CreateReply::default().embed(error_embed("You cannot poke yourself")))
            .await?;
        return Ok(());
    }

    let gif_url = get_kawaii_gif("poke", ctx.data().config.kawaii_api_key.as_ref()).await?;

    let mut embed = serenity::CreateEmbed::default()
        .description(format!("**{}** pokes **{}** 👉", ctx.author().name, member.name))
        .color(serenity::Color::from_rgb(243, 156, 18));

    if let Some(url) = gif_url {
        embed = embed.image(url);
    }

    ctx.send(poise::CreateReply::default().embed(embed)).await?;
    Ok(())
}

#[poise::command(slash_command, category = "Fun")]
pub async fn bite(
    ctx: Context<'_>,
    #[description = "User to bite"] member: serenity::User,
) -> Result<(), Error> {
    if member.id == ctx.author().id {
        ctx.send(poise::CreateReply::default().embed(error_embed("You cannot bite yourself")))
            .await?;
        return Ok(());
    }

    let gif_url = get_kawaii_gif("bite", ctx.data().config.kawaii_api_key.as_ref()).await?;

    let mut embed = serenity::CreateEmbed::default()
        .description(format!("**{}** bites **{}** 🦷", ctx.author().name, member.name))
        .color(serenity::Color::from_rgb(155, 89, 182));

    if let Some(url) = gif_url {
        embed = embed.image(url);
    }

    ctx.send(poise::CreateReply::default().embed(embed)).await?;
    Ok(())
}

#[poise::command(slash_command, category = "Fun")]
pub async fn lick(
    ctx: Context<'_>,
    #[description = "User to lick"] member: serenity::User,
) -> Result<(), Error> {
    if member.id == ctx.author().id {
        ctx.send(poise::CreateReply::default().embed(error_embed("You cannot lick yourself")))
            .await?;
        return Ok(());
    }

    let gif_url = get_kawaii_gif("lick", ctx.data().config.kawaii_api_key.as_ref()).await?;

    let mut embed = serenity::CreateEmbed::default()
        .description(format!("**{}** licks **{}** 👅", ctx.author().name, member.name))
        .color(serenity::Color::from_rgb(233, 30, 99));

    if let Some(url) = gif_url {
        embed = embed.image(url);
    }

    ctx.send(poise::CreateReply::default().embed(embed)).await?;
    Ok(())
}

#[poise::command(slash_command, category = "Fun")]
pub async fn tickle(
    ctx: Context<'_>,
    #[description = "User to tickle"] member: serenity::User,
) -> Result<(), Error> {
    if member.id == ctx.author().id {
        ctx.send(poise::CreateReply::default().embed(error_embed("You cannot tickle yourself")))
            .await?;
        return Ok(());
    }

    let gif_url = get_kawaii_gif("tickle", ctx.data().config.kawaii_api_key.as_ref()).await?;

    let mut embed = serenity::CreateEmbed::default()
        .description(format!("**{}** tickles **{}** 🤭", ctx.author().name, member.name))
        .color(serenity::Color::from_rgb(255, 215, 0));

    if let Some(url) = gif_url {
        embed = embed.image(url);
    }

    ctx.send(poise::CreateReply::default().embed(embed)).await?;
    Ok(())
}

#[poise::command(slash_command, rename = "8ball", category = "Fun")]
pub async fn eightball(
    ctx: Context<'_>,
    #[description = "Your question"] question: String,
) -> Result<(), Error> {
    let responses = [
        "It is certain.",
        "Without a doubt.",
        "Yes, definitely.",
        "You can count on it.",
        "From what I see, yes.",
        "Very likely.",
        "Outlook is good.",
        "Yes.",
        "Signs point to yes.",
        "Reply hazy, try again.",
        "Ask again later.",
        "Better not tell you now.",
        "Cannot predict now.",
        "Concentrate and ask again.",
        "Don't count on it.",
        "My answer is no.",
        "My sources say no.",
        "Outlook is not so good.",
        "Very doubtful.",
    ];

    let response = {
        let mut rng = rand::thread_rng();
        responses[rng.gen_range(0..responses.len())]
    };

    let embed = serenity::CreateEmbed::default()
        .title("🎱 Magic 8-Ball")
        .field("Question", question, false)
        .field("Answer", response, false)
        .color(serenity::Color::from_rgb(52, 152, 219))
        .footer(serenity::CreateEmbedFooter::new(format!("Requested by {}", ctx.author().name))
            .icon_url(ctx.author().face()));

    ctx.send(poise::CreateReply::default().embed(embed)).await?;
    Ok(())
}

#[poise::command(slash_command, category = "Fun")]
pub async fn joke(ctx: Context<'_>) -> Result<(), Error> {
    let jokes = [
        "Why don't scientists trust atoms? Because they make up everything.",
        "Why did the programmer quit his job? Because he didn't get arrays.",
        "Why did the scarecrow win an award? He was outstanding in his field.",
        "What do you call fake spaghetti? An impasta.",
        "Why did the math book look sad? It had too many problems.",
        "What do you call cheese that isn't yours? Nacho cheese.",
        "Why can't your nose be 12 inches long? Because then it would be a foot.",
        "Why did the bicycle fall over? It was two tired.",
        "What do you call a pile of cats? A meowtain.",
        "Why did the golfer bring two pairs of pants? In case he got a hole in one.",
    ];

    let joke = {
        let mut rng = rand::thread_rng();
        jokes[rng.gen_range(0..jokes.len())]
    };

    let embed = serenity::CreateEmbed::default()
        .title("😂 Joke")
        .description(joke)
        .color(serenity::Color::from_rgb(255, 215, 0))
        .footer(serenity::CreateEmbedFooter::new(format!("Requested by {}", ctx.author().name))
            .icon_url(ctx.author().face()));

    ctx.send(poise::CreateReply::default().embed(embed)).await?;
    Ok(())
}

#[poise::command(slash_command, category = "Fun")]
pub async fn meme(ctx: Context<'_>) -> Result<(), Error> {
    let client = reqwest::Client::new();
    let response = client
        .get("https://meme-api.com/gimme")
        .send()
        .await?;

    if !response.status().is_success() {
        ctx.send(poise::CreateReply::default().embed(error_embed("Unable to fetch a meme")))
            .await?;
        return Ok(());
    }

    let data: MemeResponse = response.json().await?;

    let color = {
        let mut rng = rand::thread_rng();
        serenity::Color::from_rgb(rng.gen(), rng.gen(), rng.gen())
    };

    let embed = serenity::CreateEmbed::default()
        .title(data.title)
        .url(data.post_link)
        .image(data.url)
        .footer(serenity::CreateEmbedFooter::new(format!("👍 {} | r/{}", data.ups, data.subreddit)))
        .color(color);

    ctx.send(poise::CreateReply::default().embed(embed)).await?;
    Ok(())
}
